transport/mux/establisher.go

```diff
@@ -0,0 +1,85 @@
+package mux
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"math/rand/v2"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/yamux"
+	"go.temporal.io/server/common/backoff"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+
+	"github.com/temporalio/s2s-proxy/config"
+	"github.com/temporalio/s2s-proxy/encryption"
+	"github.com/temporalio/s2s-proxy/metrics"
+)
+
+type establishingConnProvider struct {
+	serverAddress  string
+	tlsWrapper     func(net.Conn) net.Conn
+	logger         log.Logger
+	isShuttingDown func() bool
+	metricLabels   []string
+}
+
+// NewMuxEstablisherProvider makes an outbound call using the provided TCP settings. This constructor handles unpacking
+// the TLS config, configures the connection provider with retry and exponential backoff, and sets a disconnect
+// sleep time of 1-2 seconds.
+func NewMuxEstablisherProvider(name string, transportFn SetTransportCallback, setting config.TCPClientSetting, metricLabels []string, logger log.Logger, shutDown context.Context) (*MuxProvider, error) {
+	tlsWrapper := func(conn net.Conn) net.Conn { return conn }
+	if tlsCfg := setting.TLS; tlsCfg.IsEnabled() {
+		tlsConfig, err := encryption.GetClientTLSConfig(tlsCfg)
+		if err != nil {
+			return nil, err
+		}
+		if tlsConfig == nil {
+			return nil, errors.New("tls config was nil even though TLS is enabled")
+		}
+		tlsWrapper = func(conn net.Conn) net.Conn { return tls.Client(conn, tlsConfig) }
+	}
+	connPv := &establishingConnProvider{setting.ServerAddress, tlsWrapper, logger, func() bool { return isDone(shutDown) }, metricLabels}
+	sessionFn := func(conn net.Conn) (*yamux.Session, error) { return yamux.Client(conn, nil) }
+	disconnectFn := func() {
+		// If the server rapidly disconnects us, we don't want to get caught in a tight loop. Sleep 1-2 seconds before retry
+		time.Sleep(time.Second + time.Duration(rand.IntN(1000))*time.Millisecond)
+	}
+	return &MuxProvider{name, connPv, sessionFn, disconnectFn, transportFn, metricLabels, logger, shutDown, sync.Once{}}, nil
+}
+
+// GetConnection makes a TCP call to establish a connection, then returns it. Retries with backoff over 30 seconds
+func (p *establishingConnProvider) GetConnection() (net.Conn, error) {
+	var client net.Conn
+
+	dialFn := func() error {
+		var err error
+		client, err = net.DialTimeout("tcp", p.serverAddress, 5*time.Second)
+		if err != nil {
+			return err
+		}
+		client = p.tlsWrapper(client)
+		return nil
+	}
+
+	retryPolicy := backoff.NewExponentialRetryPolicy(time.Second).
```

Comment:
nit: I was wondering if we need to create a new retry policy each time. I think using a top-level var for this is okay (if we prefer) because when we call `backoff.ThrottleRetry` it should start from zero attempts (or from currentAttempt=1 it looks like).